python/rmm/rmm/statistics.py

```diff
@@ -0,0 +1,309 @@
+
+    if push_statistics() is None:
+        # Save the current non-statistics memory resource for later cleanup
+        prior_non_stats_mr = rmm.mr.get_current_device_resource()
+        enable_statistics()
+    else:
+        prior_non_stats_mr = None
+
+    try:
+        current_mr = rmm.mr.get_current_device_resource()
+        yield
+    finally:
+        if current_mr is not rmm.mr.get_current_device_resource():
+            raise ValueError(
+                "RMM memory source stack was changed "
+                "while in the statistics context"
+            )
+        if prior_non_stats_mr is None:
+            pop_statistics()
+        else:
+            rmm.mr.set_current_device_resource(prior_non_stats_mr)
+
+
+class ProfilerRecords:
+    """Records of the memory statistics recorded by a profiler"""
+
+    @dataclass
+    class Data:
+        """Single record of memory statistics"""
+
+        num_calls: int = 0
+        memory_total: int = 0
+        memory_peak: int = 0
+
+        def add(self, memory_total: int, memory_peak: int):
+            self.num_calls += 1
+            self.memory_total += memory_total
+            self.memory_peak = max(self.memory_peak, memory_peak)
+
+    def __init__(self) -> None:
+        self._lock = threading.Lock()
+        self._records: Dict[str, ProfilerRecords.Data] = defaultdict(
+            ProfilerRecords.Data
+        )
+
+    def add(self, name: str, data: Statistics) -> None:
+        """Add memory statistics to the record named `name`
+
+        This method is thread-safe.
+
+        Parameters
+        ----------
+        name
+            Name of the record
+        data
+            Memory statistics of `name`
+        """
+        with self._lock:
+            self._records[name].add(
+                memory_total=data.total_bytes, memory_peak=data.peak_bytes
+            )
+
+    @property
+    def records(self) -> Dict[str, Data]:
+        """Dictionary mapping record names to their memory statistics"""
+        return dict(self._records)
+
+    def pretty_print(
+        self,
+        ordered_by: Literal[
+            "num_calls", "memory_peak", "memory_total"
+        ] = "memory_peak",
+    ) -> str:
+        """Pretty format the recorded memory statistics
+
+        Parameters
+        ----------
+        ordered_by
+            Sort the statistics by this attribute.
+
+        Return
+        ------
+        The pretty formatted string of the memory statistics
+        """
+
+        # Sort by `ordered_by`
+        records = sorted(
+            ((name, data) for name, data in self.records.items()),
+            key=lambda x: getattr(x[1], ordered_by),
+            reverse=True,
+        )
+        ret = "Memory Profiling\n"
+        ret += "================\n\n"
+        if len(records) == 0:
+            return ret + "No data, maybe profiling wasn't enabled?"
+        ret += f"Ordered by: {ordered_by}\n\n"
+        ret += "ncalls     memory_peak    memory_total  "
```

Comment:
I think this line should have the units used, e.g. `... memory_peak (MiB) ... `.